### PR TITLE
implement team leaderboard functionality

### PR DESCRIPTION
**Pull request by AndrewCLu**
- [x] Define types and proving for team leaderboard
- [x] Create team leaderboard quest in dev console
- [x] Update team leaderboard scores with proving and verification
- [x] Fetch team leaderboard scores

### CODE REVIEW

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 1]
```diff
@@ -33,6 +38,57 @@ export default function QuestDetailPage() {
 
   if (!quest) return null;
 
+  const onUpdateTeamLeaderboardScore = async () => {
```
Comment (kalidiagne) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:41]: I move this part later in a hook file

File: apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      if (!areAllBigIntsDifferent(consumedSigNullifiers)) {
+        return res
+          .status(500)
+          .json({ message: "Do not provide multiple of the same signature" });
+      }
+
+      const teamLeaderboardProof = JSON.parse(
+        serializedProof
+      ) as TeamLeaderboardProof;
+      const teamCardPubKeyIndex = cardPubKeys.findIndex(
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts:108]: nit: other proofs all operate over indices and not raw pubkeys. why don't we do the same? not important though, can be refactored later

File: apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+      const consumedSigNullifiersHex = consumedSigNullifiers.map(bigIntToHex);
+      const previouslyUsedSigNullifiers =
+        await prisma.teamLeaderboardNullifiers.findMany({
+          where: {
+            questId: Number(questId),
+          },
+        });
+
+      const newlyConsumedSigNullifiers = consumedSigNullifiersHex.filter(
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts:123]: nit: Can probably do something with sets here that's much more efficient. Just add all of the `newlyConsumedSigNullifiers` to a set form of `previouslyUsedSigNullifiers` and check if they exist yet

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 4]
```diff
@@ -62,8 +118,14 @@ export default function QuestDetailPage() {
             </Link>
           );
         })}
-        <LeaderBoard></LeaderBoard>
-        <Button variant="secondary">Update team score</Button>
+        {quest.proofType === $Enums.ProofType.TEAM_LEADERBOARD && (
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:121]: Nice, this feels clean. So only quests that have team leaderboard will show a leaderboard, and their power will also be related to teams? Like do we have some sort of way of funneling the top 3 teams into a claimable power?

File: packages/jubmoji-api/src/proofs.ts [hunk 5]
```diff
@@ -323,3 +328,130 @@ export class PublicMessageSignature
...
+      sigNullifierRandomness: hexToBigInt(this.sigNullifierRandomness),
+      pubKeyNullifierRandomness,
+      pathToCircuits: this.pathToCircuits,
+    });
+
+    const sigs = [];
+    const msgHashes = [];
+    const publicInputs = [];
+    const indices = [];
+    for (const jubmoji of collectionJubmojis) {
```
Comment (vb7401) [inline packages/jubmoji-api/src/proofs.ts:384]: Are we tracking an issue where we can avoid recomputing proofs for ones we've already done? / Building some sort of localStorage stash

**Review by vb7401 — APPROVED** [review]
Overall, LGTM to merge. Some nits + issue tracking comments for later!

File: packages/jubmoji-api/src/proofs.ts (hunk 5 above)
Comment (AndrewCLu) [reply packages/jubmoji-api/src/proofs.ts:384]: https://github.com/jubmoji/jubmoji.quest/issues/67

File: apps/jubmoji-quest/src/pages/quests/[id].tsx (hunk 4 above)
Comment (AndrewCLu) [reply apps/jubmoji-quest/src/pages/quests/[id].tsx:121]: I mean, you could do this entirely on the frontend, just based on the fetched leaderboard if the user has a team pubkey in the top 3 the power is displayed

File: apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts (hunk 3 above)
Comment (AndrewCLu) [reply apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts:123]: Oops, good catch

File: apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts (hunk 2 above)
Comment (AndrewCLu) [reply apps/jubmoji-quest/src/pages/api/team-leaderboard/index.ts:108]: Thanks, we should try to keep definitions consistent, updated

**Pull request merged**